receiver/stefreceiver/stef.go

```diff
@@ -0,0 +1,165 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package stefreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/stefreceiver"
+import (
+	"context"
+	"errors"
+	"net"
+	"sync"
+
+	stefgrpc "github.com/splunk/stef/go/grpc"
+	"github.com/splunk/stef/go/grpc/stef_proto"
+	"github.com/splunk/stef/go/otel/oteltef"
+	stefpdatametrics "github.com/splunk/stef/go/pdata/metrics"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/component/componentstatus"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/receiver"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/stefreceiver/internal"
+)
+
+type stefReceiver struct {
+	cfg        *Config
+	serverGRPC *grpc.Server
+
+	nextMetrics consumer.Metrics
+	shutdownWG  sync.WaitGroup
+
+	settings receiver.Settings
+}
+
+// Start runs the STEF gRPC receiver.
+func (r *stefReceiver) Start(_ context.Context, host component.Host) error {
+	var err error
+	if r.serverGRPC, err = r.cfg.ServerConfig.ToServer(
+		context.Background(), host, r.settings.TelemetrySettings,
+	); err != nil {
+		return err
+	}
+
+	r.serverGRPC = grpc.NewServer()
+
+	r.settings.Logger.Info("Starting GRPC server", zap.String("endpoint", r.cfg.ServerConfig.NetAddr.Endpoint))
+	var gln net.Listener
+	if gln, err = r.cfg.ServerConfig.NetAddr.Listen(context.Background()); err != nil {
+		return err
+	}
+
+	schema, err := oteltef.MetricsWireSchema()
+	if err != nil {
+		return err
+	}
+
+	settings := stefgrpc.ServerSettings{
+		Logger:       nil,
+		ServerSchema: &schema,
+		MaxDictBytes: 0,
+		Callbacks:    stefgrpc.Callbacks{OnStream: r.onStream},
+	}
+
+	stefSrv := stefgrpc.NewStreamServer(settings)
+	stef_proto.RegisterSTEFDestinationServer(r.serverGRPC, stefSrv)
+
+	r.shutdownWG.Add(1)
+	go func() {
+		defer r.shutdownWG.Done()
+		if errGrpc := r.serverGRPC.Serve(gln); errGrpc != nil && !errors.Is(errGrpc, grpc.ErrServerStopped) {
+			componentstatus.ReportStatus(host, componentstatus.NewFatalErrorEvent(errGrpc))
+		}
+	}()
+	return nil
+}
+
+// Shutdown is a method to turn off receiving.
+func (r *stefReceiver) Shutdown(_ context.Context) error {
+	if r.serverGRPC != nil {
+		r.serverGRPC.GracefulStop()
+	}
+
+	r.shutdownWG.Wait()
+	return nil
+}
+
+func (r *stefReceiver) onStream(grpcReader stefgrpc.GrpcReader, stream stefgrpc.STEFStream) error {
+	r.settings.Logger.Info("Incoming STEF/gRPC connection.")
+
+	reader, err := oteltef.NewMetricsReader(grpcReader)
+	if err != nil {
+		r.settings.Logger.Error("Cannot decode data on incoming STEF/gRPC connection", zap.Error(err))
+		return err
+	}
+
+	// Create a responder for this stream and run it in a separate goroutine.
+	resp := internal.NewResponder(r.settings.Logger, stream)
+	defer resp.Stop()
+	go resp.Run()
```

Comment:
Using `errgroup.Group` here?

Reply:
that didn't really work out, but maybe a different way to build this interaction would make sense